retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java

```diff
@@ -0,0 +1,179 @@
+ *     <li>{@code httpStatus} - {@link HttpStatus#code()}</li>
+ * </ul>
+ */
+public class RetrofitClassAwareMeterIdPrefixFunction implements MeterIdPrefixFunction {
+    private static final List<Class> retrofitAnnotations = ImmutableList.of(
+            POST.class, PUT.class, PATCH.class, HEAD.class, GET.class, OPTIONS.class, HTTP.class, DELETE.class
+    );
+    private final Map<String, List<Tag>> methodNameToTags;
+    private final String name;
+    private final String serviceName;
+    private final String serviceTagName;
+
+    /**
+     * Returns a newly created {@link RetrofitClassAwareMeterIdPrefixFunction} with the specified {@code name}
+     * and {@code serviceClass}.
+     */
+    public static MeterIdPrefixFunction of(String name, Class serviceClass) {
+        return builder(name, serviceClass).build();
+    }
+
+    /**
+     * Returns a newly created {@link RetrofitMeterIdPrefixFunctionBuilder} with the specified {@code name}
+     * and {@code serviceClass}.
+     */
+    public static RetrofitMeterIdPrefixFunctionBuilder builder(String name, Class serviceClass) {
+        return new RetrofitMeterIdPrefixFunctionBuilder(requireNonNull(name, "name"))
+                .withServiceClass(serviceClass);
+    }
+
+    RetrofitClassAwareMeterIdPrefixFunction(String name,
+                                            Class serviceClass,
+                                            String serviceTagName) {
+        this.name = name;
+        this.serviceTagName = firstNonNull(serviceTagName, "service");
+        this.serviceName = serviceClass.getSimpleName();
+        this.methodNameToTags = defineTagsForMethods(serviceClass);
+    }
+
+    @Override
+    public MeterIdPrefix activeRequestPrefix(MeterRegistry registry, RequestLog log) {
+        final ImmutableList.Builder<Tag> tagsListBuilder = ImmutableList.builder();
+        buildTags(tagsListBuilder, log);
+        return new MeterIdPrefix(name, tagsListBuilder.build());
+    }
+
+    @Override
+    public MeterIdPrefix apply(MeterRegistry registry, RequestLog log) {
+        final ImmutableList.Builder<Tag> tagListBuilder = ImmutableList.builder();
+        buildTags(tagListBuilder, log);
+        RequestMetricSupport.appendHttpStatusTag(tagListBuilder, log);
+        return new MeterIdPrefix(name, tagListBuilder.build());
+    }
+
+    private void buildTags(ImmutableList.Builder<Tag> tagsBuilder, RequestLog log) {
+        final Invocation invocation = InvocationUtil.getInvocation(log);
+
+        tagsBuilder.add(Tag.of(serviceTagName, serviceName));
+        if (invocation != null) {
+            final String methodName = invocation.method().getName();
+            final List<Tag> methodTags = methodNameToTags.get(methodName);
+            if (methodTags != null) {
+                tagsBuilder.addAll(methodTags);
+                return;
+            }
+        }
+
+        tagsBuilder.add(Tag.of("method", log.method().name()));
```

Comment:
A meter must have same set of tags regardless of whether `Invocation` is available or not, because otherwise Prometheus will complain.